internal/controller/progressive/progressive.go

```diff
@@ -296,3 +329,124 @@ func IsNumaflowChildReady(upgradingObjectStatus *kubernetes.GenericStatus) bool
+  - rolloutObject: The rollout object containing the current rollout status.
+  - liveRolloutObject: The live rollout object representing the current state.
+  - existingPromotedChild: The unstructured object of the promoted child to be scaled down.
+  - skipPromotedChildSourceVerticesScaling: A flag to skip scaling if set to true.
+  - controller: The progressive controller handling the scaling logic.
+  - c: The Kubernetes client for interacting with the cluster.
+
+Returns:
+  - A boolean indicating if scaling was performed.
+  - An error if any issues occur during the scaling process.
+*/
+func scaleDownPromotedChild(
+	ctx context.Context,
+	rolloutObject ctlrcommon.RolloutObject,
+	liveRolloutObject ctlrcommon.RolloutObject,
+	existingPromotedChild *unstructured.Unstructured,
+	skipPromotedChildSourceVerticesScaling bool,
+	controller progressiveController,
+	c client.Client,
+) (bool, error) {
+
+	if skipPromotedChildSourceVerticesScaling {
+		return false, nil
+	}
+
+	if !liveRolloutObject.GetRolloutStatus().ProgressiveStatus.PromotedChildStatus.AreAllSourceVerticesScaledDown(existingPromotedChild.GetName()) {
+
+		// ScaleDownPromotedChildSourceVertices either updates the existingPromotedChild to scale down the source vertices pods or
+		// retrieves the currently running pods to update the scaleValuesMap used on the rollout status.
+		// This serves to make sure that the pods for each vertex have been really scaled down before proceeding with the progressive update.
+		scaleValuesMap, promotedChildNeedsUpdate, err := controller.ScaleDownPromotedChildSourceVertices(ctx, liveRolloutObject, existingPromotedChild, c)
+		if err != nil {
+			return false, fmt.Errorf("error updating scaling properties to the existing promoted child definition: %w", err)
+		}
+
+		if promotedChildNeedsUpdate {
+			if err = kubernetes.UpdateResource(ctx, c, existingPromotedChild); err != nil {
+				return false, fmt.Errorf("error scaling down the existing promoted child: %w", err)
+			}
+		}
+
+		if rolloutObject.GetRolloutStatus().ProgressiveStatus.PromotedChildStatus == nil {
+			rolloutObject.GetRolloutStatus().ProgressiveStatus.PromotedChildStatus = &apiv1.PromotedChildStatus{}
+		}
+		rolloutObject.GetRolloutStatus().ProgressiveStatus.PromotedChildStatus.Name = existingPromotedChild.GetName()
+		rolloutObject.GetRolloutStatus().ProgressiveStatus.PromotedChildStatus.ScaleValues = scaleValuesMap
+		rolloutObject.GetRolloutStatus().ProgressiveStatus.PromotedChildStatus.MarkAllSourceVerticesScaledDown()
+
+		// Set ScaleValuesRestoredToDesired to false in case previously set to true and now scaling back down to recover from a previous failure
+		rolloutObject.GetRolloutStatus().ProgressiveStatus.PromotedChildStatus.ScaleValuesRestoredToDesired = false
+
+		// Return that scaling down was performed
+		return true, nil
+	}
+
+	// Return that scaling down was NOT performed
+	return false, nil
+}
```

Comment:
I do like the elegance of having the `scaleDownPromotedChild()` and `scaleUpPromotedChild()` functions by the way